fix: gotouch does not detect Git repositories in Azure DevOps

prskr opened the pull request:
# What does this PR do?

instead of checking whether a URL has a `.git` suffix, send an HTTP request to check whether the URL is an HTTP Git URL. If the request fails/does not return the expected status code and content type it's assumed to be an archive URL

Potential future improvement could be to make use of the mime type detection of the `http` package to determine whether the content is actually an archive and if so of what type.

#63

denizgursoy commented:
I think that everything related to git should be under the clone package. The cloner interface should have the IsGit method that you wrote and we should call that method only in the structure task. Therefore, the logic if a url contains a git repository stay in cloner only. What do you think?

baez90 commented:
Initially I wanted to keep the impact 'as low as possible' although I admit that adding `context.Context` to like a 100 functions is quite the opposite of 'low footprint' 😅 

I think it depends on whether you want to support other version control systems as well at some point?
If so it could also make sense to add a new package `vcs` with a `Detector` interface that returns an 'enum':

- `None`
- `Git`
- ... can be extended at some point

Adding `IsGit()` to the `Cloner` feels a little bit non-Go-ish, if you know what I mean? Most Go interfaces I'm aware of are focused on a single operation and therefore only contains operations **necessary** for this very operation whereas `IsGit()` and `CloneFromUrl()` belong somewhat in the same topic but do not really belong together?

Which is why I could think of the following options:
- adding a `VCSDetector` interface to the `cloner` package
- adding a new `vcs` package and the aforementioned `Detector` interface

in both cases I'd add a new mock for the interface to be able to mock it easier than what I did so far and refactor the remaining code as you suggested.

What do you think?

denizgursoy commented:
I prefer VCSDetector interface to the cloner package. I think it would fit better

baez90 commented:
Alright, will try to refactor the current implementation and update the PR in the upcoming days 😊

denizgursoy commented:
@prskr  I want to merge this if you do the changes. I want to update dependency version and do not want to have conflicts with you

prskr commented:
I added the corresponding interface and updated the tests, I hope I did not miss anything 😅  tbh. it's been so much time, I somewhat forgot what exactly I planned to do, so I tried to implement it as easy as possible. Let me know what you think and if necessary, I'll update it during this week 😊
